### PR TITLE
Fix type errors introduced by new annotations in the Prometheus Client library.

### PR DESCRIPTION
**Pull request by reivilibre**
This also depends on https://github.com/prometheus/client_python/pull/754 which I considered to be a legitimate mistake in the type annotations.
The rest of it is likely to be fair enough and our fault, so I've made the rest of the fixes necessary.

### CODE REVIEW

**Review by DMRobertson — APPROVED** [review]
SGTM.

What, if anything, goes wrong at runtime if we pass in a label value that's not a string? Do we lose metrics data? Does the runtime automatically convert the value to a string?

**Comment by DMRobertson** [comment]
I'm going to merge in develop to make sure this passes typechecking with the most recent changes.

**Comment by reivilibre** [comment]
> What, if anything, goes wrong at runtime if we pass in a label value that's not a string? Do we lose metrics data? **Does the runtime automatically convert the value to a string?**

Yeah, it has a built-in `str()` call to convert it.

**Comment by reivilibre** [comment]
@DMRobertson 
This should now do the trick as of 0.13.1 of prometheus-client (with thanks to the maintainer!).
I've reverted your version cap in the requirements file, plus I've rolled back the change to `str()` all the labels since that's no longer needed.

File: synapse/python_dependencies.py [hunk 1]
```diff
@@ -76,8 +76,7 @@
     "msgpack>=0.5.2",
     "phonenumbers>=8.2.0",
     # we use GaugeHistogramMetric, which was added in prom-client 0.4.0.
-    # 0.13.0 has an incorrect type annotation, see #11832.
-    "prometheus_client>=0.4.0,<0.13.0",
+    "prometheus_client>=0.4.0",
```
[the remark this answers is not in the document]
Comment (reivilibre) [reply synapse/python_dependencies.py:79]: I also think not — in the extreme, it'd end up with us maintaining a list of versions that don't work.

**Comment by darix** [comment]
just curious ... would 0.13.1 be a good release of the prometheus client to be used with the 1.52.0 release?

**Pull request merged**